Mathlib/Control/Functor.lean

```diff
@@ -0,0 +1,302 @@
+/-
+Copyright (c) 2017 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Simon Hudon
+-/
+import Mathlib.Control.Basic
+import Mathlib.Init.Set
+
+/-!
+# Functors
+
+This module provides additional lemmas, definitions, and instances for `Functor`s.
+
+## Main definitions
+
+* `Functor.Const α` is the functor that sends all types to `α`.
+* `Functor.AddConst α` is `Functor.Const α` but for when `α` has an additive structure.
+* `Functor.Comp F G` for functors `F` and `G` is the functor composition of `F` and `G`.
+* `Liftp` and `Liftr` respectively lift predicates and relations on a type `α`
+  to `F α`.  Terms of `F α` are considered to, in some sense, contain values of type `α`.
+
+## Tags
+
+functor, applicative
+-/
+
+
+attribute [functor_norm] seq_assoc pure_seq map_pure seq_map_assoc map_seq
+
+universe u v w
+
+section Functor
+
+variable {F : Type u → Type v}
+
+variable {α β γ : Type u}
+
+variable [Functor F] [LawfulFunctor F]
+
+theorem Functor.map_id : (· <$> ·) id = (id : F α → F α) := by apply funext <;> apply id_map
+#align functor.map_id Functor.map_id
+
+theorem Functor.map_comp_map (f : α → β) (g : β → γ) :
+    ((· <$> ·) g ∘ (· <$> ·) f : F α → F γ) = (· <$> ·) (g ∘ f) :=
+  funext <| fun _ => (comp_map _ _ _).symm
+  -- porting note: was `apply funext <;> intro <;> sorry --rw [comp_map]` but `rw` failed?
+#align functor.map_comp_map Functor.map_comp_map
+
+theorem Functor.ext {F} :
+    ∀ {F1 : Functor F} {F2 : Functor F} [@LawfulFunctor F F1] [@LawfulFunctor F F2]
+    (_H : ∀ (α β) (f : α → β) (x : F α), @Functor.map _ F1 _ _ f x = @Functor.map _ F2 _ _ f x),
+    F1 = F2
+  | ⟨m, mc⟩, ⟨m', mc'⟩, H1, H2, H => by
+    cases show @m = @m' by funext α β f x; apply H
+    -- porting note: Lean marks `H` as an unused variable?
```

Comment:
inaccessible, not unused. This is a hygiene issue. You can probably use `cases ... with ...` to make the name accessible.

Reply:
No, note that `H` is a hypothesis. It marked it as unused before I added the underscore.

Reply:
Oh, I see. That's actually perfectly reasonable. It's just looking at the forall in the return type, where indeed H is not used.